File: tutorials/tour-of-restate-rust/src/part1/checkout_service.rs
use std::collections::HashSet;
use restate_sdk::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CheckoutRequest {
    pub(crate) user_id: String,
    pub(crate) tickets: HashSet<String>,
}

// <start_checkout>
#[restate_sdk::service]
pub(crate) trait CheckoutService {
    async fn handle(request: Json<CheckoutRequest>) -> Result<bool, HandlerError>;
}

pub struct CheckoutServiceImpl;

impl CheckoutService for CheckoutServiceImpl {
    async fn handle(
        &self,
        _ctx: Context<'_>,
        Json(CheckoutRequest { user_id: _, tickets: _ }): Json<CheckoutRequest>,
    ) -> Result<bool, HandlerError> {
        Ok(true)
    }
}
// <end_checkout>

File: go/tutorials/tour-of-restate-go/part3/checkoutservice.go
package main

import restate "github.com/restatedev/sdk-go"

// <start_checkout>
type CheckoutService struct{}

type CheckoutRequest struct {
	UserId  string   `json:"userId"`
	Tickets []string `json:"tickets"`
}

func (CheckoutService) Handle(ctx restate.Context, request CheckoutRequest) (bool, error) {
	return true, nil
}

// <end_checkout>
